yellowbrick/classifier/class_balance_heat_map.py

```diff
@@ -0,0 +1,169 @@
+# yellowbrick.classifier.class_balance_heat_map
+# Class balance visualizer for showing per-class support.
+#
+# Author:   Rebecca Bilbro <rbilbro@districtdatalabs.com>
+# Author:   Benjamin Bengfort <bbengfort@districtdatalabs.com>
+# Author:   Neal Humphrey
+# Author:   Larry Gray
+# Created:  Sun March 4 20:45:00 2018 -0400
+#
+# Copyright (C) 2017 District Data Labs
+# For license information, see LICENSE.txt
+#
+# ID: class_balance_heat_map.py [5388065] lwgray@gmail.com $
+
+"""
+Class balance heat map visualizer for showing per-class support.
+"""
+
+##########################################################################
+## Imports
+##########################################################################
+
+from matplotlib import cm
+import matplotlib.pyplot as plt
+import numpy as np
+
+from .base import ClassificationScoreVisualizer
+
+from sklearn.model_selection import train_test_split
+from sklearn.utils.multiclass import unique_labels
+
+
+##########################################################################
+## Class Balance Heat Map Chart
+##########################################################################
+
+class ClassBalanceHeatMap(ClassificationScoreVisualizer):
+    """
+    Class balance heat map that shows the support for each class in the
+    fitted classification model displayed as a stacked bar.  Each bar is
+    segmented to show the distribution of predicted classes for each
+    class. It is initialized with a fitted model and generates a
+    class balance heat map chart on draw.
+
+    Parameters
+    ----------
+    ax: axes
+        the axis to plot the figure on.
+    model: estimator
+        Scikit-Learn estimator object. Should be an instance of a classifier,
+        else ``__init__()`` will raise an exception.
+    classes: list
+        A list of class names for the legend. If classes is None and a y value
+        is passed to fit then the classes are selected from the target vector.
+    kwargs: dict
+        Keyword arguments passed to the super class. Here, used
+        to colorize the bars in the histogram.
+    Notes
+    -----
+    These parameters can be influenced later on in the visualization
+    process, but can and should be set as early as possible.
+    """
+
+    def score(self, X, y=None, **kwargs):
+        """
+        Generates a 2D array where each row is the count of the
+        predicted classes and each column is the true class
+
+        Parameters
+        ----------
+        X : ndarray or DataFrame of shape n x m
+            A matrix of n instances with m features
+        y : ndarray or Series of length n
+            An array or series of target or class values
+
+        Returns
+        -------
+
+        ax : the axis with the plotted figure
+        """
+
+        y_pred = self.predict(X)
+        self.indices = unique_labels(y, y_pred)
+        self.classes_ = self.classes_ if self.classes_ is not None else self.indices
```

Comment:
I would advise not storing indices on `self` -- changing this line to:

```
indices = unique_labels(y, y_pred)
if self.classes_ is None:
    self.classes_ = indices 
```

Estimator attributes should either be hyperparameters (supplied by the user and having reasonable defaults) or parameters, things that are added during `fit` and are [suffixed by an underscore](http://scikit-learn.org/stable/tutorial/statistical_inference/settings.html#estimators-objects). 

However, I'm not sure if we actually do need to do anything about `self.classes_` - in what cases would we fit and `self.classes_` is None? Instead, this should probably be a check to align indices with `self.classes_` -- if there is a mismatch we need to raise an error ourselves rather than raising an uninterpretable error. (This would be the case where more labels are supplied in our test data, via score than in our training data; if there are less labels in y and y_pred, that's fine - we just need to ensure there are zeros for the fit classes that do not have any associated values). 

I'd also recommend we add the above two cases to the tests for this.

Reply:
1. I was not sure what exception to raise, but I choose to go with ModelError
2. Also, I wan't sure what message to to raise with Error

Reply:
@bbengfort if i don't store indices on self, the the scope of the indices variable is limited to the score function and can't be accesses by draw() Is there a work around for this?

Reply:
@bbengfort disregard my last comment, you addressed it below

Reply:
scikit-learn does some target verification in [`_check_targets`](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/metrics/classification.py#L44) 

Interestingly it does not necessarily do what I suggested above, but perhaps it is sufficient to just run `y` and `y_pred` through that check. They raise `ValueError` when things are off, I would prefer to raise `MantleValueError` if possible.

Reply:
@bbengfort What is MantleValueError?

Reply:
Whoops, sorry - [`YellowbrickValueError`](https://github.com/DistrictDataLabs/yellowbrick/blob/develop/yellowbrick/exceptions.py#L56) - mantle is a project at my work.